sqld/src/http/services/idle_shutdown.rs

```diff
@@ -0,0 +1,74 @@
+use std::sync::Arc;
+use std::time::Duration;
+
+use tokio::sync::{watch, Notify};
+use tokio::time::timeout;
+use tower::{Layer, Service};
+
+pub struct IdleShutdownLayer {
+    watcher: Arc<watch::Sender<()>>,
+}
+
+impl IdleShutdownLayer {
+    pub fn new(idle_timeout: Duration, shutdown_notifier: Arc<Notify>) -> Self {
+        let (sender, mut receiver) = watch::channel(());
+        tokio::spawn(async move {
+            loop {
+                // FIXME: if we measure that this is causing performance issues, we may want to
+                // implement some debouncing.
+                let timeout_fut = timeout(idle_timeout, receiver.changed());
+                match timeout_fut.await {
+                    Ok(Ok(_)) => continue,
+                    Ok(Err(_)) => break,
```

Comment:
Is it guaranteed that the server will shutdown when hit this branch? I'm trying to assess if it would be possible to hit this branch and the instance not shutting down. That would mean a working instance that never scales down, no? Unless I'm misunderstanding something.

Maybe it makes sense to call `shutdown_notifier.notify_waiters()` here as well?

Reply:
or not break out of the loop at least

Reply:
This branch is hit when the sender has be dropped on the other side, so we won't be receiving messages anymore. I think it's safe to assume that the server is already shutting down in this case. (It could also be a hard reset)

Reply:
👍 thanks for the explanation. Would it make sense to to at least log a warning in case we ever have to debug somewhere around here?